pkg/controller/acmechallenges/sync.go

```diff
@@ -222,8 +222,25 @@ func (c *Controller) syncChallengeStatus(ctx context.Context, cl acmecl.Interfac
 		return err
 	}
 
-	// TODO: should we validate the State returned by the ACME server here?
+	switch acmeChallenge.Status {
+	case acmeapi.StatusPending, acmeapi.StatusValid, acmeapi.StatusInvalid:
+		// do nothing
+	default:
+		// bogus status from acme API that we can't handle.
+		return fmt.Errorf("acme api returned bogus status for challenge: %v", acmeChallenge.Status)
+	}
```

Comment:
There are a variety of other ACME statuses that could be returned, and returning an error here will only cause the sync to be retried after applying a back-off.

Instead, could we just set the `ch.Status.Reason` if `acmeChallenge.Error.Detail != ""`? Seems like any 'detail' field returned by the server would be useful to be honest.

Another example - if the ACME server has rate limited us, the 'state' of the Order (or Challenge) could remain 'pending' but the error 'detail' could still contain valuable information.

Reply:
These states are based off the comments [here](https://github.com/jetstack/cert-manager/blob/4d3179f2ba38114edc79293f4b0f541b1f029ba5/third_party/crypto/acme/types.go#L220). I figured I'd turn bogus states into errors, since they can take over the internal state machine and we'd have no idea how to handle them. Better get an error in that case.